File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/gamepad/GamepadJoystick.java
package net.kdt.pojavlaunch.customcontrols.gamepad;

import android.view.InputDevice;
import android.view.MotionEvent;

import net.kdt.pojavlaunch.utils.MathUtils;

public class GamepadJoystick {

    // 方向
    public static final int DIRECTION_NONE = -1; // 中央にあるGamepadJoystick

    public static final int DIRECTION_EAST = 0;
    public static final int DIRECTION_NORTH_EAST = 1;
    public static final int DIRECTION_NORTH = 2;
    public static final int DIRECTION_NORTH_WEST = 3;
    public static final int DIRECTION_WEST = 4;
    public static final int DIRECTION_SOUTH_WEST = 5;
    public static final int DIRECTION_SOUTH = 6;
    public static final int DIRECTION_SOUTH_EAST = 7;

    private final InputDevice mInputDevice;

    private final int mHorizontalAxis;
    private final int mVerticalAxis;
    private float mVerticalAxisValue = 0;
    private float mHorizontalAxisValue = 0;
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public GamepadJoystick(int horizontalAxis, int verticalAxis, InputDevice device){
        mHorizontalAxis = horizontalAxis;
        mVerticalAxis = verticalAxis;
        this.mInputDevice = device;
    }
/**
 * 「AngleRadian」の値を取得します。
 */
    public double getAngleRadian(){
        // -PIからPIまで
        // TODO misuse of the deadzone here !
        return -Math.atan2(getVerticalAxis(), getHorizontalAxis());
    }
/**
 * 「AngleDegree」の値を取得します。
 */
    public double getAngleDegree(){
        // 0から360度まで
        double result = Math.toDegrees(getAngleRadian());
        if(result < 0) result += 360;

        return result;
    }
/**
 * 「Magnitude」の値を取得します。
 */
    public double getMagnitude(){
        float x = Math.abs(mHorizontalAxisValue);
        float y = Math.abs(mVerticalAxisValue);

        return MathUtils.dist(0,0, x, y);
    }
/**
 * 「VerticalAxis」の値を取得します。
 */
    public float getVerticalAxis(){
        return mVerticalAxisValue;
    }
/**
 * 「HorizontalAxis」の値を取得します。
 */
    public float getHorizontalAxis(){
        return mHorizontalAxisValue;
    }
/**
 * このオブジェクトが「JoystickEvent」状態であるかを判定します。
 */
    public static boolean isJoystickEvent(MotionEvent event){
        return (event.getSource() & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK
                && event.getAction() == MotionEvent.ACTION_MOVE;
    }
/**
 * 「HeightDirection」の値を取得します。
 */
    public int getHeightDirection(){
        if(getMagnitude() == 0) return DIRECTION_NONE;
        return ((int) ((getAngleDegree()+22.5)/45)) % 8;
    }
/**
 * 「XAxisValue」の値を設定します。
 */


    /* Setters */
    public void setXAxisValue(float value){
        this.mHorizontalAxisValue = value;
    }
/**
 * 「YAxisValue」の値を設定します。
 */
    public void setYAxisValue(float value){
        this.mVerticalAxisValue = value;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/gamepad/GamepadJoystick.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.gamepad

import android.view.InputDevice
import android.view.MotionEvent
import com.arata.yukarilauncher.utils.MathUtils

/**
 * ゲームパッドのジョイスティック（アナログスティック）を表すクラス。
 * 水平/垂直軸の値から角度、大きさ、方向を計算します。
 */
class GamepadJoystick {
    companion object {
        /** 無方向。 */
        const val DIRECTION_NONE = -1
        /** 東（右）。 */
        const val DIRECTION_EAST = 0
        /** 北東。 */
        const val DIRECTION_NORTH_EAST = 1
        /** 北（上）。 */
        const val DIRECTION_NORTH = 2
        /** 北西。 */
        const val DIRECTION_NORTH_WEST = 3
        /** 西（左）。 */
        const val DIRECTION_WEST = 4
        /** 南西。 */
        const val DIRECTION_SOUTH_WEST = 5
        /** 南（下）。 */
        const val DIRECTION_SOUTH = 6
        /** 南東。 */
        const val DIRECTION_SOUTH_EAST = 7

        /**
         * MotionEventがジョイスティックイベントかどうかを判定します。
         * @param event 判定するMotionEvent
         * @return ジョイスティックイベントの場合はtrue
         */
        fun isJoystickEvent(event: MotionEvent): Boolean {
            return (event.source and InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK
                    && event.action == MotionEvent.ACTION_MOVE
        }
    }

    private val mInputDevice: InputDevice
    private val mHorizontalAxis: Int
    private val mVerticalAxis: Int
    private var mVerticalAxisValue = 0f
    private var mHorizontalAxisValue = 0f

    /**
     * @param horizontalAxis 水平軸の定数
     * @param verticalAxis 垂直軸の定数
     * @param device 入力デバイス
     */
    constructor(horizontalAxis: Int, verticalAxis: Int, device: InputDevice) {
        mHorizontalAxis = horizontalAxis
        mVerticalAxis = verticalAxis
        this.mInputDevice = device
    }

    /** @return ジョイスティックの角度（ラジアン） */
    val angleRadian: Double
        get() = -Math.atan2(getVerticalAxis().toDouble(), getHorizontalAxis().toDouble())

    /** @return ジョイスティックの角度（度） */
    val angleDegree: Double
        get() {
            var result = Math.toDegrees(angleRadian)
            if (result < 0) result += 360.0
            return result
        }

    /** @return ジョイスティックの変位の大きさ */
    val magnitude: Double
        get() {
            val x = Math.abs(mHorizontalAxisValue)
            val y = Math.abs(mVerticalAxisValue)
            return MathUtils.dist(0f, 0f, x, y).toDouble()
        }

    /** @return 垂直軸の値 */
    fun getVerticalAxis(): Float = mVerticalAxisValue

    /** @return 水平軸の値 */
    fun getHorizontalAxis(): Float = mHorizontalAxisValue

    /** @return 現在の方向を示す8方向の定数 */
    val heightDirection: Int
        get() {
            if (magnitude == 0.0) return DIRECTION_NONE
            return ((angleDegree + 22.5) / 45).toInt() % 8
        }

    /** @param value X軸の値を設定 */
    fun setXAxisValue(value: Float) {
        this.mHorizontalAxisValue = value
    }

    /** @param value Y軸の値を設定 */
    fun setYAxisValue(value: Float) {
        this.mVerticalAxisValue = value
    }
}